qemu: add user space emulator only variant

jcaesar opened the pull request:
## Description of changes

This adds a variant of qemu that builds in pkgsStatic. The goal is using it as a binfmt interpreter that works across mount namespaces (e.g. enabling `docker run --platform …`).

This PR takes out only the changes to `pkgs/` from #300070, which itself is a fix and resubmission of #160802. Additionally, it makes the following changes:
* Stop referring to `pkgsStatic`
* Rename `qemu-user-static` to `qemu-user`
* Stop depending on texinfo when not building docs, which avoids running into #295608, thus avoiding the dependency on #299623
* (rebase on master, fixing a simple merge conflict)


You can convince yourself that this works by starting a VM
```bash
nix run -vL --expr 'let
  nixpkgs = builtins.getFlake "github:NixOS/nixpkgs/d57f30155eb628f27f12d24c3e1fd6a30ee7fee7";
  sys = nixpkgs.lib.nixosSystem {
    system = "x86_64-linux";
    modules = [
      "${nixpkgs}/nixos/modules/virtualisation/qemu-vm.nix"
      (
        { pkgs, ... }:
        {
          services.getty.autologinUser = "root";
          virtualisation.docker.enable = true;
          virtualisation.graphics = false;
          boot.binfmt.emulatedSystems = ["aarch64-linux"];
          boot.binfmt.registrations.aarch64-linux = {
            interpreter = "${pkgs.pkgsStatic.qemu-user}/bin/qemu-aarch64";
            wrapInterpreterInShell = false;
            fixBinary = true;
            openBinary = true; # debatable, see https://github.com/NixOS/nixpkgs/pull/314998#issuecomment-2237347334
            matchCredentials = true;
            preserveArgvZero = true;
          };
        }
      )
    ];
  };
in sys.config.system.build.vm'
```
and then running
```bash
docker run --rm --platform linux/arm64  alpine uname -m
```
inside of it. (`docker run` as a demonstration. The real benefit is being able to build multi-platform docker images on one host.)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

zhaofengli committed:
qemu: Skip doc-related sphinx dependencies sphinx when docs are disabled

zhaofengli committed:
qemu: Support disabling tools

Not needed for a minimal qemu-user-static build.

zhaofengli committed:
qemu: Support disabling TCG plugins

The plugin .so's do not build statically.

zhaofengli committed:
qemu: Support disabling blobs

Not needed for a minimal qemu-user-static build.

jcaesar commented:
cot (Python 3.12) and virtualbox  builds fail, but that seems to be the same on master. If there's a correct way of naming this kind of PR, please do tell.

---

Result of `nixpkgs-review pr 314998` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ocamlPackages.ocaml-freestanding</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_6_9.virtualbox</li>
    <li>linuxKernel.packages.linux_6_9_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>python312Packages.cot</li>
    <li>python312Packages.cot.dist</li>
  </ul>
</details>
<details>
  <summary>79 packages built:</summary>
  <ul>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>alpine-make-vm-image</li>
    <li>cloud-init</li>
    <li>cloud-init.dist</li>
    <li>cloud-utils</li>
    <li>cloud-utils.guest</li>
    <li>colima</li>
    <li>cot (python311Packages.cot)</li>
    <li>cot.dist (python311Packages.cot.dist)</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>gnome.gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>lima</li>
    <li>lima-bin</li>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_6_6_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_8.virtualbox</li>
    <li>linuxKernel.packages.linux_6_8_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_stable.virtualbox)</li>
    <li>lxd-lts</li>
    <li>multipass</li>
    <li>nemu</li>
    <li>open-watcom-bin</li>
    <li>open-watcom-bin-unwrapped</li>
    <li>out-of-tree</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>qemu</li>
    <li>qemu-user</li>
    <li>qemu-user.debug</li>
    <li>qemu-utils</li>
    <li>qemu-utils.debug</li>
    <li>qemu.debug</li>
    <li>qemu.ga</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qemu_kvm</li>
    <li>qemu_kvm.debug</li>
    <li>qemu_kvm.ga</li>
    <li>qemu_test</li>
    <li>qemu_test.debug</li>
    <li>qemu_test.ga</li>
    <li>qtemu</li>
    <li>quickemu</li>
    <li>quickgui</li>
    <li>solo5</li>
    <li>solo5.debug</li>
    <li>vagrant</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
    <li>zpool-auto-expand-partitions</li>
  </ul>
</details>

TECHNOFAB11 commented:
Tried this and it works great btw!
Will try this on my servers to see if it works in Kubernetes aswell, but as podman works it should work with containerd aswell.

```nix
# input:
    nixpkgs-qemu-patch.url = "github:jcaesar/fork2pr-nixpkgs/pr-6";

# config:
    binfmt.emulatedSystems = ["aarch64-linux"];
    binfmt.registrations.aarch64-linux = let
      pkgs_patch = import inputs.nixpkgs-qemu-patch {
        inherit system;
      };
    in {
      interpreter = "${pkgs_patch.pkgsStatic.qemu-user}/bin/qemu-aarch64";
      wrapInterpreterInShell = false;
      fixBinary = true;
      openBinary = true;
      matchCredentials = true;
      preserveArgvZero = true;
    };
```

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -33892,6 +33892,9 @@ with pkgs;
   };
 
+  # variant of qemu building user space emulator only - intended to be used from pkgsStatic
+  qemu-user = callPackage ../applications/virtualization/qemu/user.nix {};
```
This should be inlined like the other variants, otherwise overriding won't work.

jcaesar replied:
Done. (Avoided spamming `all-packages.nix` with 30 lines of overrides by introducing a new option as you [previously suggested](pkgs/applications/virtualization/qemu/default.nix).)

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -136,7 +142,13 @@ stdenv.mkDerivation (finalAttrs: {
-  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch;
+  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch
+
+  ## FIXME: libaio does not provide a pkg-info file;
```
What's a pkg-info file.  Do you mean pkg-config?  Can we fix that?

alyssais reviewed (changes requested):
Needs to be rebased.  There shouldn't be a commit like b7c0c9f8bbec639d93f6b48fb52fbb8db30f253c — updates that need to be made should just be folded into the appropriate commits.

I'd like to do a more thorough review when I have time and https://github.com/NixOS/nixpkgs/pull/299623 is done.

jcaesar commented:
> Needs to be rebased. There shouldn't be a commit like b7c0c9f — updates that need to be made should just be folded into the appropriate commits.

I decided to take this a step further and squash all commits and rewrote the commit message accordingly, since they didn't made sense in isolation. Do tell if you want me to undo this.

> I'd like to do a more thorough review when I have time and #299623 is done.

Fortunately, #299623 is no longer a dependency for building this PR (i.e. building `pkgsStatic.qemu-user`). You can safely ignore it. (I guess the commit messages didn't make this clear, sorry.)  
Please give this a look when you find the time and motivation.

> This should be inlined like the other variants, otherwise overriding won't work.

Done. (Avoided spamming `all-packages.nix` with 30 lines of overrides by introducing a new option as you [previously suggested](pkgs/applications/virtualization/qemu/default.nix).)

> What's a pkg-info file. Do you mean pkg-config? Can we fix that?

Probably pkg-info, yes. Fixing might be hard, but it turns out libaio is only(?) used for block drive emulation, and thus not necessary for qemu-user. Conditionally disabled libaio, the patch is gone.

(Looking into this a bit, libaio is a plain Makefile build, and it doesn't have a maintainer. Not difficult to add a pkg-config, but not pretty either. On the other hand side, qemu's meson build uses `find_library`, which doesn't seem to work with pkg-config files anyway(?).)

---

<details><summary>This caused all kinds of changes, which are hard to understand with force-push + rebase. Here's the diff between "after rebase" and "now":</summary>

```patch
diff --git a/pkgs/applications/virtualization/qemu/aio-find-static-library.patch b/pkgs/applications/virtualization/qemu/aio-find-static-library.patch
deleted file mode 100644
index 5faf0693dc10..000000000000
--- a/pkgs/applications/virtualization/qemu/aio-find-static-library.patch
+++ /dev/null
@@ -1,26 +0,0 @@
-diff --git a/meson.build b/meson.build
-index c9c3217ba4..1a8de03f4e 100644
---- a/meson.build
-+++ b/meson.build
-@@ -986,7 +986,8 @@ zlib = dependency('zlib', required: true)
- libaio = not_found
- if not get_option('linux_aio').auto() or have_block
-   libaio = cc.find_library('aio', has_headers: ['libaio.h'],
--                           required: get_option('linux_aio'))
-+                           required: get_option('linux_aio'),
-+                           dirs: [get_option('linux_aio_path')])
- endif
- 
- linux_io_uring_test = '''
-diff --git a/meson_options.txt b/meson_options.txt
-index 0a99a059ec..84d2449de9 100644
---- a/meson_options.txt
-+++ b/meson_options.txt
-@@ -186,6 +186,7 @@ option('libusb', type : 'feature', value : 'auto',
-        description: 'libusb support for USB passthrough')
- option('linux_aio', type : 'feature', value : 'auto',
-        description: 'Linux AIO support')
-+option('linux_aio_path', type: 'string', value: '', description: 'Path for libaio.a')
- option('linux_io_uring', type : 'feature', value : 'auto',
-        description: 'Linux io_uring support')
- option('lzfse', type : 'feature', value : 'auto',
diff --git a/pkgs/applications/virtualization/qemu/default.nix b/pkgs/applications/virtualization/qemu/default.nix
index 82c59d0ad52c..92c9d0d56a8b 100644
--- a/pkgs/applications/virtualization/qemu/default.nix
+++ b/pkgs/applications/virtualization/qemu/default.nix
@@ -5,45 +5,53 @@
 , makeWrapper, removeReferencesTo
 , attr, libcap, libcap_ng, socat, libslirp
 , CoreServices, Cocoa, Hypervisor, rez, setfile, vmnet
-, guestAgentSupport ? (with stdenv.hostPlatform; isLinux || isNetBSD || isOpenBSD || isSunOS || isWindows) && !toolsOnly
-, numaSupport ? stdenv.isLinux && !stdenv.isAarch32 && !toolsOnly, numactl
-, seccompSupport ? stdenv.isLinux && !toolsOnly, libseccomp
-, alsaSupport ? lib.hasSuffix "linux" stdenv.hostPlatform.system && !nixosTestRunner && !toolsOnly
-, pulseSupport ? !stdenv.isDarwin && !nixosTestRunner && !toolsOnly, libpulseaudio
-, pipewireSupport ? !stdenv.isDarwin && !nixosTestRunner && !toolsOnly, pipewire
-, sdlSupport ? !stdenv.isDarwin && !nixosTestRunner && !toolsOnly, SDL2, SDL2_image
-, jackSupport ? !stdenv.isDarwin && !nixosTestRunner && !toolsOnly, libjack2
-, gtkSupport ? !stdenv.isDarwin && !xenSupport && !nixosTestRunner && !toolsOnly, gtk3, gettext, vte, wrapGAppsHook3
-, vncSupport ? !nixosTestRunner && !toolsOnly, libjpeg, libpng
-, smartcardSupport ? !nixosTestRunner && !toolsOnly, libcacard
-, spiceSupport ? true && !nixosTestRunner && !toolsOnly, spice, spice-protocol
-, ncursesSupport ? !nixosTestRunner && !toolsOnly, ncurses
+, guestAgentSupport ? (with stdenv.hostPlatform; isLinux || isNetBSD || isOpenBSD || isSunOS || isWindows) && !minimal
+, numaSupport ? stdenv.isLinux && !stdenv.isAarch32 && !minimal, numactl
+, seccompSupport ? stdenv.isLinux && !minimal, libseccomp
+, alsaSupport ? lib.hasSuffix "linux" stdenv.hostPlatform.system && !nixosTestRunner && !minimal
+, pulseSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, libpulseaudio
+, pipewireSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, pipewire
+, sdlSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, SDL2, SDL2_image
+, jackSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, libjack2
+, gtkSupport ? !stdenv.isDarwin && !xenSupport && !nixosTestRunner && !minimal, gtk3, gettext, vte, wrapGAppsHook3
+, vncSupport ? !nixosTestRunner && !minimal, libjpeg, libpng
+, smartcardSupport ? !nixosTestRunner && !minimal, libcacard
+, spiceSupport ? true && !nixosTestRunner && !minimal, spice, spice-protocol
+, ncursesSupport ? !nixosTestRunner && !minimal, ncurses
 , usbredirSupport ? spiceSupport, usbredir
 , xenSupport ? false, xen
 , cephSupport ? false, ceph
 , glusterfsSupport ? false, glusterfs, libuuid
 , openGLSupport ? sdlSupport, mesa, libepoxy, libdrm
-, rutabagaSupport ? openGLSupport && !toolsOnly && lib.meta.availableOn stdenv.hostPlatform rutabaga_gfx, rutabaga_gfx
+, rutabagaSupport ? openGLSupport && !minimal && lib.meta.availableOn stdenv.hostPlatform rutabaga_gfx, rutabaga_gfx
 , virglSupport ? openGLSupport, virglrenderer
-, libiscsiSupport ? !toolsOnly, libiscsi
+, libiscsiSupport ? !minimal, libiscsi
 , smbdSupport ? false, samba
-, tpmSupport ? !toolsOnly
-, uringSupport ? stdenv.isLinux, liburing
-, canokeySupport ? !toolsOnly, canokey-qemu
-, capstoneSupport ? !toolsOnly, capstone
+, tpmSupport ? !minimal
+, uringSupport ? stdenv.isLinux && !userOnly, liburing
+, canokeySupport ? !minimal, canokey-qemu
+, capstoneSupport ? !minimal, capstone
 , pluginsSupport ? !stdenv.hostPlatform.isStatic
-, enableDocs ? true
-, enableTools ? true
-, enableBlobs ? true
+, enableDocs ? !minimal || toolsOnly
+, enableTools ? !minimal || toolsOnly
+, enableBlobs ? !minimal || toolsOnly
 , hostCpuOnly ? false
 , hostCpuTargets ? (if toolsOnly
                     then [ ]
                     else if hostCpuOnly
                     then (lib.optional stdenv.isx86_64 "i386-softmmu"
                           ++ ["${stdenv.hostPlatform.qemuArch}-softmmu"])
-                    else null)
+                    else if userOnly
+                    then map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-linux-user") [
+                      "aarch64-linux" "armv7l-linux"
+                      "i386-linux" "x86_64-linux"
+                      "powerpc-linux" "powerpc64-linux" "powerpc64le-linux"
+                      "riscv32-linux" "riscv64-linux"
+                    ] else null)
 , nixosTestRunner ? false
 , toolsOnly ? false
+, userOnly ? false
+, minimal ? toolsOnly || userOnly
 , gitUpdater
 , qemu-utils # for tests attribute
 }:
@@ -70,21 +78,21 @@ stdenv.mkDerivation (finalAttrs: {
 
   nativeBuildInputs = [
     makeWrapper removeReferencesTo
-    pkg-config flex bison dtc meson ninja
+    pkg-config flex bison meson ninja
 
     # Don't change this to python3 and python3.pkgs.*, breaks cross-compilation
     python3Packages.python
   ]
     ++ lib.optionals gtkSupport [ wrapGAppsHook3 ]
-    ++ lib.optionals enableDocs [ python3Packages.sphinx python3Packages.sphinx-rtd-theme ]
+    ++ lib.optionals enableDocs [ texinfo python3Packages.sphinx python3Packages.sphinx-rtd-theme ]
     ++ lib.optionals hexagonSupport [ glib ]
-    ++ lib.optionals stdenv.isDarwin [ sigtool ];
+    ++ lib.optionals stdenv.isDarwin [ sigtool ]
+    ++ lib.optionals (!userOnly) [ dtc ];
 
-  buildInputs = [ dtc zlib glib pixman
+  buildInputs = [ zlib glib pixman
     vde2 lzo snappy libtasn1
     gnutls nettle curl libslirp
   ]
-    ++ lib.optionals enableDocs [ texinfo ]
     ++ lib.optionals ncursesSupport [ ncurses ]
     ++ lib.optionals stdenv.isDarwin [ CoreServices Cocoa Hypervisor rez setfile vmnet ]
     ++ lib.optionals seccompSupport [ libseccomp ]
@@ -99,7 +107,8 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optionals smartcardSupport [ libcacard ]
     ++ lib.optionals spiceSupport [ spice-protocol spice ]
     ++ lib.optionals usbredirSupport [ usbredir ]
-    ++ lib.optionals stdenv.isLinux [ libaio libcap_ng libcap attr ]
+    ++ lib.optionals stdenv.isLinux [ libcap_ng libcap attr ]
+    ++ lib.optionals (stdenv.isLinux && !userOnly) [ libaio ]
     ++ lib.optionals xenSupport [ xen ]
     ++ lib.optionals cephSupport [ ceph ]
     ++ lib.optionals glusterfsSupport [ glusterfs libuuid ]
@@ -110,7 +119,8 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optionals smbdSupport [ samba ]
     ++ lib.optionals uringSupport [ liburing ]
     ++ lib.optionals canokeySupport [ canokey-qemu ]
-    ++ lib.optionals capstoneSupport [ capstone ];
+    ++ lib.optionals capstoneSupport [ capstone ]
+    ++ lib.optionals (!userOnly) [ dtc ];
 
   dontUseMesonConfigure = true; # meson's configurePhase isn't compatible with qemu build
   dontAddStaticConfigureFlags = true;
@@ -142,13 +152,7 @@ stdenv.mkDerivation (finalAttrs: {
       revert = true;
     })
   ]
-  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch
-
-  ## FIXME: libaio does not provide a pkg-info file;
-  # and meson does not support static libraries lookup path using -L, relying on LIBRARY_PATH (https://github.com/mesonbuild/meson/issues/10172);
-  # and musl-gcc does not support LIBRARY_PATH overrides (https://www.openwall.com/lists/musl/2017/02/22/7);
-  # so we have to patch the meson.build to add the libaio path to the search path manually.
-  ++ lib.optional stdenv.hostPlatform.isStatic ./aio-find-static-library.patch;
+  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch;
 
   postPatch = ''
     # Otherwise tries to ensure /var/run exists.
@@ -184,7 +188,7 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optional usbredirSupport "--enable-usb-redir"
     ++ lib.optional (hostCpuTargets != null) "--target-list=${lib.concatStringsSep "," hostCpuTargets}"
     ++ lib.optionals stdenv.isDarwin [ "--enable-cocoa" "--enable-hvf" ]
-    ++ lib.optional stdenv.isLinux "--enable-linux-aio"
+    ++ lib.optional (stdenv.isLinux && !userOnly) "--enable-linux-aio"
     ++ lib.optional gtkSupport "--enable-gtk"
     ++ lib.optional xenSupport "--enable-xen"
     ++ lib.optional cephSupport "--enable-rbd"
@@ -199,12 +203,7 @@ stdenv.mkDerivation (finalAttrs: {
     ++ lib.optional capstoneSupport "--enable-capstone"
     ++ lib.optional (!pluginsSupport) "--disable-plugins"
     ++ lib.optional (!enableBlobs) "--disable-install-blobs"
-    ++ lib.optionals stdenv.hostPlatform.isStatic [
-      "--static"
-      # FIXME: "multiple definition of `strtoll'" with libnbcompat
-      "--extra-ldflags=-Wl,--allow-multiple-definition"
-      "-Dlinux_aio_path=${libaio}/lib"
-    ];
+    ++ lib.optional stdenv.hostPlatform.isStatic "--static";
 
   dontWrapGApps = true;
 
@@ -228,6 +227,11 @@ stdenv.mkDerivation (finalAttrs: {
     for f in $out/bin/qemu-system-*; do
       wrapGApp $f
     done
+  '' + lib.optionalString stdenv.hostPlatform.isStatic ''
+    # HACK: Otherwise the result will have the entire buildinput closure
+    # injected by the pkgsStatic stdenv
+    # <https://github.com/NixOS/nixpkgs/issues/83667>
+    rm -f $out/nix-support/propagated-build-inputs
   '';
   preBuild = "cd build";
 
@@ -267,10 +271,8 @@ stdenv.mkDerivation (finalAttrs: {
   '';
 
   # Add a ‘qemu-kvm’ wrapper for compatibility/convenience.
-  postInstall = lib.optionalString (!toolsOnly) ''
-    if [ -f $out/bin/qemu-system-${stdenv.hostPlatform.qemuArch} ]; then
-      ln -s $out/bin/qemu-system-${stdenv.hostPlatform.qemuArch} $out/bin/qemu-kvm
-    fi
+  postInstall = lib.optionalString (!minimal) ''
+    ln -s $out/bin/qemu-system-${stdenv.hostPlatform.qemuArch} $out/bin/qemu-kvm
   '';
 
   passthru = {
@@ -298,7 +300,13 @@ stdenv.mkDerivation (finalAttrs: {
     platforms = platforms.unix;
   }
   # toolsOnly: Does not have qemu-kvm and there's no main support tool
-  // lib.optionalAttrs (!toolsOnly) {
+  # userOnly: There's one qemu-<arch> for every architecture
+  // lib.optionalAttrs (!toolsOnly && !userOnly) {
     mainProgram = "qemu-kvm";
+  }
+  # userOnly: Only meaningful on linux
+  // lib.optionalAttrs userOnly {
+    platforms = lib.platforms.linux;
+    description = "QEMU User space emulator - launch executables compiled for one CPU on another CPU";
   };
 })
diff --git a/pkgs/applications/virtualization/qemu/user.nix b/pkgs/applications/virtualization/qemu/user.nix
deleted file mode 100644
index eeb9199ec6d4..000000000000
--- a/pkgs/applications/virtualization/qemu/user.nix
+++ /dev/null
@@ -1,63 +0,0 @@
-# When used from pkgsStatic, this provides qemu-user-static
-{
-  lib,
-  qemu,
-}:
-
-let
-
-  qemuUserPlatforms = [
-    "aarch64-linux"
-    "armv7l-linux"
-    "i386-linux"
-    "powerpc-linux"
-    "powerpc64-linux"
-    "powerpc64le-linux"
-    "riscv32-linux"
-    "riscv64-linux"
-    "x86_64-linux"
-  ];
-
-  qemuTargets = map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-linux-user") qemuUserPlatforms;
-
-  qemuUser = (qemu.override {
-    guestAgentSupport = false;
-    numaSupport = false;
-    seccompSupport = false;
-    alsaSupport = false;
-    pulseSupport = false;
-    sdlSupport = false;
-    jackSupport = false;
-    pipewireSupport = false;
-    gtkSupport = false;
-    vncSupport = false;
-    smartcardSupport = false;
-    spiceSupport = false;
-    ncursesSupport = false;
-    libiscsiSupport = false;
-    smbdSupport = false;
-    tpmSupport = false;
-    uringSupport = false;
-    capstoneSupport = false;
-    enableDocs = false;
-    enableTools = false;
-    enableBlobs = false;
-    hostCpuTargets = qemuTargets;
-  }).overrideAttrs (old: {
-    # HACK: Otherwise the result will have the entire buildinput closure
-    # injected by the pkgsStatic stdenv
-    # <https://github.com/NixOS/nixpkgs/issues/83667>
-    postFixup = (old.postFixup or "") + ''
-      rm -f $out/nix-support/propagated-build-inputs
-    '';
-
-    meta = old.meta // {
-      platforms = lib.platforms.linux;
-      description = "QEMU User space emulator - launch executables compiled for one CPU on another CPU";
-    };
-  });
-in qemuUser // {
-  passthru = (qemuUser.passthru or {}) // {
-    inherit qemuUserPlatforms;
-  };
-}
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 9331dcac3146..ee5901d479d1 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -33863,7 +33863,9 @@ with pkgs;
   };
 
   # variant of qemu building user space emulator only - intended to be used from pkgsStatic
-  qemu-user = callPackage ../applications/virtualization/qemu/user.nix {};
+  qemu-user = qemu.override {
+    userOnly = true;
+  };
 
   canokey-qemu = callPackage ../applications/virtualization/qemu/canokey-qemu.nix { };
 
```

</details>

alyssais commented:
> I decided to take this a step further and squash all commits and rewrote the commit message accordingly, since they didn't made sense in isolation.

Oh, I liked all the separate commits.  They made reviewing easier.

jcaesar committed:
qemu: add workaround for static build closure pulling in buildinput (#83667)

jcaesar commented:
> Oh, I liked all the separate commits. They made reviewing easier.

There you go. (I had to reorganize things a bit because the newly added flags and disabled dependencies mean that some of the commits in the middle stopped making sense.)

Cryolitia reviewed (approved):
Result of `nixpkgs-review pr 314998` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages marked as broken and skipped:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>qemu_xen</li>
    <li>qemu_xen-light</li>
    <li>qemu_xen-light.debug</li>
    <li>qemu_xen-light.ga</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.ga</li>
    <li>qemu_xen_4_15</li>
    <li>qemu_xen_4_15-light</li>
    <li>qemu_xen_4_15-light.debug</li>
    <li>qemu_xen_4_15-light.ga</li>
    <li>qemu_xen_4_15.debug</li>
    <li>qemu_xen_4_15.ga</li>
  </ul>
</details>
<details>
  <summary>35 packages failed to build:</summary>
  <ul>
    <li>cot (python312Packages.cot)</li>
    <li>cot.dist (python312Packages.cot.dist)</li>
    <li>lima-bin</li>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxPackages_4_19_hardened.virtualbox (linuxKernel.packages.linux_4_19_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxPackages_5_10_hardened.virtualbox (linuxKernel.packages.linux_5_10_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxPackages_5_15_hardened.virtualbox (linuxKernel.packages.linux_5_15_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxPackages_5_4_hardened.virtualbox (linuxKernel.packages.linux_5_4_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxPackages_6_1_hardened.virtualbox (linuxKernel.packages.linux_6_1_hardened.virtualbox)</li>
    <li>linuxPackages.virtualbox (linuxKernel.packages.linux_6_6.virtualbox)</li>
    <li>linuxPackages_hardened.virtualbox (linuxPackages_6_6_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_6_8.virtualbox</li>
    <li>linuxPackages_6_8_hardened.virtualbox (linuxKernel.packages.linux_6_8_hardened.virtualbox)</li>
    <li>linuxPackages_latest.virtualbox (linuxKernel.packages.linux_6_9.virtualbox)</li>
    <li>linuxPackages_6_9_hardened.virtualbox (linuxKernel.packages.linux_6_9_hardened.virtualbox)</li>
    <li>linuxPackages_latest-libre.virtualbox (linuxKernel.packages.linux_latest_libre.virtualbox)</li>
    <li>linuxPackages-libre.virtualbox (linuxKernel.packages.linux_libre.virtualbox)</li>
    <li>linuxPackages_lqx.virtualbox (linuxKernel.packages.linux_lqx.virtualbox)</li>
    <li>linuxPackages_xanmod.virtualbox (linuxKernel.packages.linux_xanmod.virtualbox)</li>
    <li>linuxPackages_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_latest.virtualbox ,linuxPackages_xanmod_stable.virtualbox)</li>
    <li>linuxPackages_zen.virtualbox (linuxKernel.packages.linux_zen.virtualbox)</li>
    <li>nemu</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
  </ul>
</details>
<details>
  <summary>50 packages built:</summary>
  <ul>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>alpine-make-vm-image</li>
    <li>cloud-init</li>
    <li>cloud-init.dist</li>
    <li>cloud-utils</li>
    <li>cloud-utils.guest</li>
    <li>colima</li>
    <li>gnome.gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>libguestfs</li>
    <li>lima</li>
    <li>lxd-lts</li>
    <li>multipass</li>
    <li>open-watcom-bin</li>
    <li>open-watcom-bin-unwrapped</li>
    <li>out-of-tree</li>
    <li>python311Packages.cot</li>
    <li>python311Packages.cot.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>qemu</li>
    <li>qemu-user</li>
    <li>qemu-user.debug</li>
    <li>qemu-utils</li>
    <li>qemu-utils.debug</li>
    <li>qemu.debug</li>
    <li>qemu.ga</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qemu_kvm</li>
    <li>qemu_kvm.debug</li>
    <li>qemu_kvm.ga</li>
    <li>qemu_test</li>
    <li>qemu_test.debug</li>
    <li>qemu_test.ga</li>
    <li>qtemu</li>
    <li>quickemu</li>
    <li>quickgui</li>
    <li>solo5</li>
    <li>solo5.debug</li>
    <li>tests.testers.lycheeLinkCheck.network</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tests.testers.runNixOSTest-example</li>
    <li>tests.trivial-builders.references</li>
    <li>vagrant</li>
    <li>zpool-auto-expand-partitions</li>
  </ul>
</details>

Tested running archlinux-riscv64 on nixos-x86_64, LGTM

Aleksanaa commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -208,6 +227,11 @@ stdenv.mkDerivation (finalAttrs: {
       wrapGApp $f
     done
+  '' + lib.optionalString stdenv.hostPlatform.isStatic ''
+    # HACK: Otherwise the result will have the entire buildinput closure
```
```suggestion
    # HACK: Otherwise the result will have the entire buildInputs closure
```

jcaesar replied:
[Ty](https://github.com/NixOS/nixpkgs/compare/775b948c91c88eee09e707a925929aa8bf66fdeb..029f986c6b9f64edb55778bd1e083860696ac2a2)

Aleksanaa reviewed (approved):
Changes lgtm. @alyssais could you give a final review?

dramforever commented:
This isn't a blocker to this PR at all since it concerns only the example in the top post but

```
            openBinary = true;
            matchCredentials = true;
```

These two break coreutils `make check` which depends on being able to close stdin / fd 0. Removing these two options seem to be all that's needed to fix the problem

(It's `openBinary` that causes the problem but `matchCredentials` requires that.)

Posting this so that people having this problem can search "coreutils" on this page and find this post.

jcaesar commented:
@dramforever Thank you, that's interesting to know. I was thinking about submitting a PR next that changes the binfmt module to make that the default, but I guess I have to think about that. 

Just FYI, not setting `matchCredentials` breaks e.g. emulated sudo:
```
sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?
```
and I'm not sure it doesn't possibly have security implications.

(Looking around, [multiarch/qemu-user-static](https://github.com/multiarch/qemu-user-static) doesn't set it, debian's qemu-user-static does. I'll have to dig.)

dramforever commented:
@jcaesar I believe I have narrowed it down to a QEMU bug: https://gitlab.com/qemu-project/qemu/-/issues/2448 so there's a good chance `openBinary` will be fixed eventually.

Security implications of setting `matchCredentials` is still unclear though.

alyssais commented:
Hey, I've been sick for a while and still not quite better yet, but just wanted to say I'd still like to review this, and I'll prioritise doing that once I'm better. :)

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -276,7 +300,13 @@ stdenv.mkDerivation (finalAttrs: {
+  }
+  # userOnly: Only meaningful on linux
+  // lib.optionalAttrs userOnly {
+    platforms = lib.platforms.linux;
```
QEMU can also emulate BSD from BSD.

jcaesar replied:
I'm unsure what that implies. I tried building `pkgsCross.x86_64-{open,free,net}bsd.qemu-user`, but just got stuck on various missing dependencies. I could set this to `with lib.platforms; [ linux freebsd openbsd netbsd ];` or `lib.platforms.unix`, but I cant test it… I did at least update the comment.

jcaesar replied:
I see. Set to `linux ++ freebsd ++ openbsd ++ netbsd` now. I think updating the hostCpuTargets defaults to conditionally include `*-bsd-user` is not worth it, since this is a theoretic exercise.

alyssais reviewed (commented):
This is looking good!

I don't really like how complicated all the option defaults are becoming — I think it would be better to just set them in the qemu-utils/qemu-user expressions in all-packages.nix rather than having this layer of indirection and complicated expressions in the package, but that problem predates this PR so it doesn't need to be fixed here.

jcaesar commented:
> I don't really like how complicated all the option defaults are becoming

Agreed. I don't like putting them into all-packages though. I have two ideas:
1. Make qemu-user (and qemu-utils) separate packages, get rid of most of the options. Sadly, the separate definitions aren't quite as small as you might think. [Here's](https://github.com/jcaesar/sysflake/blob/42f93893b6bad1d3c39c1b75739e7d30de8be424/pkgs/qemu-user.nix) how that would look.
2. Put the overrides into a separate file and define `qemu-user = qemu.override (import ../aplications/virtualization/qemu/qemu-user-overrides.nix);` in all-packages.

But I sure wouldn't be unhappy about doing that in a separate PR.

alyssais commented:
> Agreed. I don't like putting them into all-packages though. I have two ideas:

Why not?  That's how lots of other packages do it.

>     1. Make qemu-user (and qemu-utils) separate packages, get rid of most of the options. Sadly, the separate definitions aren't quite as small as you might think. [Here's](https://github.com/jcaesar/sysflake/blob/42f93893b6bad1d3c39c1b75739e7d30de8be424/pkgs/qemu-user.nix) how that would look.

Yeah I would rather not do this.  Sharing the implementation has a lot of value.

>     2. Put the overrides into a separate file and define `qemu-user = qemu.override (import ../aplications/virtualization/qemu/qemu-user-overrides.nix);` in all-packages.

This would be fine, but it seems weird to have a special way of doing this that only QEMU does, when plenty of other packages just set the options in all-packages.nix…

alyssais commented:
I'm a bit confused by the commit history here — "qemu: Skip dependencies unnecessary for qemu-user which fail in pkgsS…" comes after "qemu: Fix static build", so does the latter commit fix static or not?

jcaesar commented:
[074bb5f](https://github.com/NixOS/nixpkgs/pull/314998/commits/074bb5fa23f6a020bc7e82e2d06f96c53d90b2ba) fixes the static build of qemu itself, but at that point you need to patch some of the dependencies (like [this](https://github.com/jcaesar/sysflake/blob/1eba2f9cb649bd46deeb86d857e8a2fb9187d091/mod/binfmt.nix#L20-L23) or [that](https://github.com/NixOS/nixpkgs/pull/299623)) to actually build it. Should I reword or reorder?

Freax13 commented:
> ... but at that point you still need to patch some of the dependencies (like [this](https://github.com/jcaesar/sysflake/blob/1eba2f9cb649bd46deeb86d857e8a2fb9187d091/mod/binfmt.nix#L20-L23) or [that](https://github.com/NixOS/nixpkgs/pull/299623)) to actually build it. Should I reword or reorder?

FYI, QEMU [doesn't need the texinfo dependency](https://gitlab.com/qemu-project/qemu/-/commit/df07c72a74fd43bfc1e6431a19eb3252a3cfe1dd) anymore. Removing the texinfo dependency also removes the implicit dependency on perl.

alyssais commented:
> Should I reword or reorder?

Reorder I think — make each commit demonstrate progress towards the ultimate goal without the need for further modifications.  Squashing these a bit is probably fine too.

jcaesar committed:
qemu: remove unused texinfo dependency

jcaesar commented:
> Reorder I think
 
Done

> QEMU [doesn't need the texinfo dependency](https://gitlab.com/qemu-project/qemu/-/commit/df07c72a74fd43bfc1e6431a19eb3252a3cfe1dd)

Also removed that.

--- 

> > > I don't really like how complicated all the option defaults are becoming

> >    Agreed. I don't like putting them into all-packages though. I have two ideas:

> Why not? That's how lots of other packages do it.

There's a lot of overrides, you'll need them twice (qemu-utils + qemu-user), and in case of hostCpuTargets, they're also fairly complex. Putting that into a 41k line file, instead of 300 line file its actually related to, seems counterproductive.

alyssais commented:
> There's a lot of overrides, you'll need them twice (qemu-utils + qemu-user), and in case of hostCpuTargets, they're also fairly complex. Putting that into a 41k line file, instead of 300 line file its actually related to, seems counterproductive.

I think the size of the file means it doesn't matter whether it ends up with a few lines of overrides, even if it's quite a lot, but duplicated complex option definitions is a fair argument.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -5,45 +5,53 @@
+                      "i386-linux" "x86_64-linux"
+                      "powerpc-linux" "powerpc64-linux" "powerpc64le-linux"
+                      "riscv32-linux" "riscv64-linux"
+                    ] else null)
```
Since this is a static list anyway, why elaborate (which IIRC is expensive), rather than listing the QEMU names directly?  They're not very likely to change.

jcaesar replied:
> which IIRC is expensive

TIL. I was just imitating the [binfmt module here](https://github.com/NixOS/nixpkgs/blob/ef6a56e11f41625d0259a86db2da07f66119f192/nixos/modules/system/boot/binfmt.nix#L32).

I poked around a bit, and from what I found, I think it's nicer to not set a target list at all (static or lib.platforms.linux) and let qemu build all `*-user` targets.

This has three advantages
* It lessens the "options are complex" problem
* If Qemu does add a target, that'll automatically be included 
* If bsd ever builds, that'll have the correct target list
But it also has two wrinkles
* The build output gets a bit large, 121 MB instead of 37
* Hexagon needs excluding, for reasons I don't fully understand

alyssais replied:
I agree that this is nicer.  I'll have a quick look at Hexagon.

alyssais replied:
Hexagon builds fine for me on aarch64-linux?

jcaesar replied:
Sorry, I should have explained this better. I [already default-disabled](https://github.com/NixOS/nixpkgs/pull/314998/commits/70ada6be384bc1ae4fa135d708936f32ac98a408#diff-f885f3c7293fd60915aa7c2d28eea89dea548af3d732700fa1224b0cafa55a0aR202-R203) hexagon on static tools only. If I don't do that, the build fails for both `.#legacyPackages.{x86_64,aarch64}-linux.pkgsStatic.qemu-user` with the following log:

```
Program flex found: YES (/nix/store/z1fkfc5ca0rra7m5zrh8i92mx18w9i00-flex-2.6.4/bin/flex)
Program bison found: YES 3.8.2 3.8.2 (/nix/store/abm55q8k9lsbz7kv2v68ns7hkdks2v2q-bison-3.8.2/bin/bison)
Found pkg-config: YES (/nix/store/zdvrzlvzbn9ymb0z8na50w995j8np16z-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Did not find CMake 'cmake'
Found CMake: NO
Build-time dependency glib-2.0 found: NO

../target/hexagon/meson.build:191:15: ERROR: Dependency lookup for glib-2.0 with method 'pkgconfig' failed: Could not generate libs for glib-2.0:
Package libpcre2-8 was not found in the pkg-config search path.
Perhaps you should add the directory containing `libpcre2-8.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libpcre2-8', required by 'glib-2.0', not found
```

There is a `dependency('glib-2.0', native: true)` in there. Which I thought should work since it's `native`. (rel: [`native` meson docs](https://github.com/mesonbuild/meson/blob/fa4f2339465ce3d755e2df802ebd5aa962e2ad27/docs/yaml/functions/dependency.yaml#L148-L151)). Hm.

alyssais replied:
oh, this is only a problem when building it statically? I think that's what I was doing wrong.

alyssais replied:
Adding pcre2 to nativeBuildInputs fixes it for me, but I think it's a bug that this is required.  Looking some more…

alyssais replied:
I've sent this [patch](https://lore.kernel.org/qemu-devel/20240805104921.4035256-1-hi@alyssa.is/) upstream that would avoid the need to add pcre2 to nativeBuildInputs.

jcaesar replied:
> Adding pcre2 to nativeBuildInputs fixes it for me, but I think it's a bug that this is required. Looking some more…

Uff, I thought I had tried that. I think that's a fine workaround for this PR.

Thank you very much for looking this deep.

jcaesar committed:
qemu: Skip dependencies unnecessary for qemu-user which fail in pkgsStatic (libaio, dtc)

zhaofengli committed:
qemu: Fix static build

alyssais reviewed (commented):
Might be nice to append something to pname for userOnly?

jcaesar committed:
qemu-user: add qemu "user space emulator only" variant

When used from pkgsStatic, this is synonymous with the "qemu-user-static"
packages available in other distros.

jcaesar commented:
> Might be nice to append something to pname for userOnly?

nix eval --raw .#legacyPackages.x86_64-linux.pkgsStatic.qemu-user.drvPath
/nix/store/ris5n3d5jd3x2vm1i4fiw0sp382qz7d9-qemu-user-static-x86_64-unknown-linux-musl-9.0.2.drv

Ah, falls into place nicely. (Did require a rebase on master though.)

alyssais reviewed (approved):
Thanks for your persistence!

jcaesar commented:
I assume we're waiting on aarch64-darwin? x86_64-darwin timed out. Should I do something about that? (Not I'd know what…)

alyssais commented:
Darwin timing out running the tests is sadly expected.  I think a lot of aarch64-darwin builds got dropped, but it's unlikely this would break on just one Darwin.
